include/revng/Model/CMakeLists.txt

```diff
@@ -0,0 +1,89 @@
+#
+
+# Header files containing model YAML must be added here
+set(MODEL_HEADERS
+    CABIFunctionType.h
+    EnumType.h
+    PrimitiveType.h
+    PrimitiveTypeKind.h
+    RawFunctionType.h
+    StructType.h
+    TypedefType.h
+    Type.h
+    TypeKind.h
+    UnionType.h
+    ABI.h
+    AbstractFunctionEdge.h
+    Architecture.h
+    Argument.h
+    BasicBlock.h
+    Binary.h
+    CallEdge.h
+    DynamicFunction.h
+    EnumEntry.h
+    FunctionEdge.h
+    FunctionEdgeType.h
+    Function.h
+    FunctionType.h
+    NamedTypedRegister.h
+    QualifiedType.h
+    Qualifier.h
+    QualifierKind.h
+    Register.h
+    RegisterState.h
+    Segment.h
+    StructField.h
+    TypedRegister.h
+    UnionField.h
+)
+
+set(MODEL_TEMPLATES
+    "${CMAKE_SOURCE_DIR}/scripts/tuple_tree_generator/tuple_tree_generator/templates/class_forward_decls.tpl.hpp"
+    "${CMAKE_SOURCE_DIR}/scripts/tuple_tree_generator/tuple_tree_generator/templates/enum.tpl.hpp"
+    "${CMAKE_SOURCE_DIR}/scripts/tuple_tree_generator/tuple_tree_generator/templates/struct.tpl.hpp"
+    "${CMAKE_SOURCE_DIR}/scripts/tuple_tree_generator/tuple_tree_generator/templates/struct_forward_decls.tpl.hpp"
+    "${CMAKE_SOURCE_DIR}/scripts/tuple_tree_generator/tuple_tree_generator/templates/struct_late.tpl.hpp"
+    "${CMAKE_SOURCE_DIR}/scripts/tuple_tree_generator/tuple_tree_generator/templates/struct_impl.tpl.cpp"
+)
+
+# -- Collect YAML from manually written headers
+set(COLLECTED_YAML_PATH "${CMAKE_BINARY_DIR}/include/revng/Model/model_schema.yml")
+
+add_custom_command(
+    OUTPUT "${COLLECTED_YAML_PATH}"
+    COMMAND "${CMAKE_SOURCE_DIR}/scripts/tuple_tree_generator/extract_yaml.py"
+    ARGS "MODEL-YAML" ${MODEL_HEADERS} > "${COLLECTED_YAML_PATH}"
+    WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+    DEPENDS ${MODEL_HEADERS}
+)
+
+# -- Generate C++ headers from collected yaml
+set(OUTPUT_DIR "${CMAKE_BINARY_DIR}/include/revng/Model/Generated")
+
+foreach(HEADER ${MODEL_HEADERS})
+  get_filename_component(HEADER_FILENAME "${HEADER}" NAME)
+  get_filename_component(HEADER_FILENAME_WE "${HEADER}" NAME_WE)
+  set(EARLY_OUTPUT "${OUTPUT_DIR}/Early/${HEADER_FILENAME}")
+  set(LATE_OUTPUT "${OUTPUT_DIR}/Late/${HEADER_FILENAME}")
+  set(IMPL_OUTPUT "${OUTPUT_DIR}/Impl/${HEADER_FILENAME_WE}.cpp")
+  list(APPEND GENERATED_HEADERS "${EARLY_OUTPUT}")
+  list(APPEND GENERATED_HEADERS "${LATE_OUTPUT}")
+  list(APPEND GENERATED_IMPLS "${IMPL_OUTPUT}")
+endforeach()
+list(APPEND GENERATED_HEADERS "${OUTPUT_DIR}/ForwardDecls.h")
+
+add_custom_command(
+    COMMAND "${CMAKE_SOURCE_DIR}/scripts/tuple_tree_generator/main.py"
+    ARGS --cpp-headers "${COLLECTED_YAML_PATH}" "${OUTPUT_DIR}"
+    OUTPUT ${GENERATED_HEADERS} ${GENERATED_IMPLS}
+    DEPENDS "${COLLECTED_YAML_PATH}" ${MODEL_TEMPLATES}
+)
+
+add_custom_target(
+    generated-model-classes
+    DEPENDS ${GENERATED_HEADERS} ${GENERATED_IMPLS}
+)
+
+revng_add_analyses_library(revngModelGenerated revng ${GENERATED_IMPLS})
```

Comment:
Why have a distinct library?
Let's put this in the `revngModel` library.

Reply:
Because `revng_add_analyses_library` must be invoked only once, it cannot add files incrementally to a given library (it's not a limitation of `revng_add_analyses_library` itself, it uses CMake builtin `add_library`, which has this limitation).
There are two workarounds as far as I know:
- having a global list of files which have to be added to the model
- moving the CMake stuff for the autogenerated model where the "main" model library is defined. I didn't do this because I found having the CMake that generates stuff from the model header files in their folder much clearer

Which option do you prefer? Personally I'm for option 2 if we must do away with the second library.

Reply:
(settled IRL for option 2)